// @tryghost/collections/build/Collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collection = void 0;
const errors_1 = require("@tryghost/errors");
const tpl_1 = __importDefault(require("@tryghost/tpl"));
const nql = require("@tryghost/nql");
const bson_objectid_1 = __importDefault(require("bson-objectid"));
const messages = {
    invalidIDProvided: 'Invalid ID provided for Collection',
    invalidDateProvided: 'Invalid date provided for {fieldName}',
    invalidFilterProvided: {
        message: 'Invalid filter provided for automatic Collection',
        context: 'Automatic type of collection should always have a filter value'
    },
    noTitleProvided: 'Title must be provided',
    slugMustBeUnique: 'Slug must be unique'
};
class Collection {
    id;
    title;
    _slug;
    get slug() {
        return this._slug;
    }
    async setSlug(slug, uniqueChecker) {
        if (slug === this.slug) {
            return;
        }
        if (await uniqueChecker.isUniqueSlug(slug)) {
            this._slug = slug;
        }
        else {
            throw new errors_1.ValidationError({
                message: (0, tpl_1.default)(messages.slugMustBeUnique)
            });
        }
    }
    description;
    type;
    filter;
    featureImage;
    createdAt;
    updatedAt;
    get deletable() {
        return this.slug !== 'index' && this.slug !== 'featured';
    }
    _deleted = false;
    _posts;
    get posts() {
        return this._posts;
    }
    get deleted() {
        return this._deleted;
    }
    set deleted(value) {
        if (this.deletable) {
            this._deleted = value;
        }
    }
    async edit(data, uniqueChecker) {
        if (this.type === 'automatic' && (data.filter === null || data.filter === '')) {
            throw new errors_1.ValidationError({
                message: (0, tpl_1.default)(messages.invalidFilterProvided.message),
                context: (0, tpl_1.default)(messages.invalidFilterProvided.context)
            });
        }
        if (data.title !== undefined) {
            this.title = data.title;
        }
        if (data.slug !== undefined) {
            await this.setSlug(data.slug, uniqueChecker);
        }
        if (data.description !== undefined) {
            this.description = data.description;
        }
        if (data.filter !== undefined) {
            this.filter = data.filter;
        }
        if (data.featureImage !== undefined) {
            this.featureImage = data.featureImage;
        }
        return this;
    }
    postMatchesFilter(post) {
        const filterNql = nql(this.filter);
        return filterNql.queryJSON(post);
    }
    /**
     * @param post {{id: string}} - The post to add to the collection
     * @param index {number} - The index to insert the post at, use negative numbers to count from the end.
     */
    addPost(post, index = -0) {
        if (this.type === 'automatic') {
            const matchesFilter = this.postMatchesFilter(post);
            if (!matchesFilter) {
                return false;
            }
        }
        if (this.posts.includes(post.id)) {
            this._posts = this.posts.filter(id => id !== post.id);
        }
        if (index < 0 || Object.is(index, -0)) {
            index = this.posts.length + index;
        }
        this.posts.splice(index, 0, post.id);
        return true;
    }
    removePost(id) {
        if (this.posts.includes(id)) {
            this._posts = this.posts.filter(postId => postId !== id);
        }
    }
    includesPost(id) {
        return this.posts.includes(id);
    }
    removeAllPosts() {
        this._posts = [];
    }
    constructor(data) {
        this.id = data.id;
        this.title = data.title;
        this._slug = data.slug;
        this.description = data.description;
        this.type = data.type;
        this.filter = data.filter;
        this.featureImage = data.featureImage;
        this.createdAt = data.createdAt;
        this.updatedAt = data.updatedAt;
        this.deleted = data.deleted;
        this._posts = data.posts;
    }
    toJSON() {
        return {
            id: this.id,
            title: this.title,
            slug: this.slug,
            description: this.description,
            type: this.type,
            filter: this.filter,
            featureImage: this.featureImage,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt,
            posts: this.posts
        };
    }
    static validateDateField(date, fieldName) {
        if (!date) {
            return new Date();
        }
        if (date instanceof Date) {
            return date;
        }
        throw new errors_1.ValidationError({
            message: (0, tpl_1.default)(messages.invalidDateProvided, { fieldName })
        });
    }
    static async create(data) {
        let id;
        if (!data.id) {
            id = new bson_objectid_1.default();
        }
        else if (typeof data.id === 'string') {
            id = bson_objectid_1.default.createFromHexString(data.id);
        }
        else if (data.id instanceof bson_objectid_1.default) {
            id = data.id;
        }
        else {
            throw new errors_1.ValidationError({
                message: (0, tpl_1.default)(messages.invalidIDProvided)
            });
        }
        if (data.type === 'automatic' && !data.filter) {
            // @NOTE: add filter validation here
            throw new errors_1.ValidationError({
                message: (0, tpl_1.default)(messages.invalidFilterProvided.message),
                context: (0, tpl_1.default)(messages.invalidFilterProvided.context)
            });
        }
        if (!data.title) {
            throw new errors_1.ValidationError({
                message: (0, tpl_1.default)(messages.noTitleProvided)
            });
        }
        return new Collection({
            id: id.toHexString(),
            title: data.title,
            slug: data.slug,
            description: data.description || null,
            type: data.type || 'manual',
            filter: data.filter || null,
            featureImage: data.feature_image || null,
            createdAt: Collection.validateDateField(data.created_at, 'created_at'),
            updatedAt: Collection.validateDateField(data.updated_at, 'updated_at'),
            deleted: data.deleted || false,
            posts: data.posts || []
        });
    }
}
exports.Collection = Collection;
